### PR TITLE
Adds static and dynamic verification for ConstantDynamic Entry

### PR DESCRIPTION
**Pull request by fengxue-IS**
- Add validation to ldc ldc_w and ldc2_w for ConstantDynamic entry
- Check that the Condy return type matches the bytecode
- Refactor ldc with more informative error messages
- Update description comments on pushLdcType(...)
- Update pushLdcType to expect a constant dynamic to provide a field
- Create new NLS entries for constantDynamic errors

Replaces: #1631
Close: #1278

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

**Comment by fengxue-IS** [comment]
@ChengJin01 can you take a look please

**Comment by fengxue-IS** [comment]
fixed code format, @ChengJin01 please take a look, thanks

File: runtime/bcverify/staticverify.c [hunk 1]
```diff
@@ -257,10 +291,22 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
...
 			tag = (UDATA) info->tag;
-			if (!((tag == CFR_CONSTANT_Double)
-				  || (tag == CFR_CONSTANT_Long))) {
-				errorType = J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT__ID;
-				/* Jazz 82615: Set the constant pool index to show up in the error message framework */
+
+			if (CFR_CONSTANT_Dynamic == tag) {
+				J9CfrConstantPoolInfo *constantDynamicSignature = &classfile->constantPool[classfile->constantPool[info->slot2].slot2];
+				if (((flags & BCT_MajorClassFileVersionMask) < BCT_Java11MajorVersionShifted)
+				&& (0 != (flags & BCT_MajorClassFileVersionMask))
```
Comment (ChengJin01) [inline runtime/bcverify/staticverify.c]: Better use J9_ARE_ANY_BITS_SET() for (0 != (flags & BCT_MajorClassFileVersionMask)

Comment (fengxue-IS) [reply runtime/bcverify/staticverify.c]: fixed

**Comment by fengxue-IS** [comment]
per discussion with @DanHeidinga, removed the zero major version check on constant dynamic as cfdumper no longer sets default classfile version to zero.

**Comment by ChengJin01** [comment]
lgtm

**Comment by fengxue-IS** [comment]
senior review: @DanHeidinga please take a look

File: runtime/bcverify/staticverify.c [hunk 2]
```diff
@@ -228,26 +228,58 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
...
-				goto _verifyError;
+			{
+				switch (tag) {
+				case CFR_CONSTANT_Integer:
+				case CFR_CONSTANT_Float:
+				case CFR_CONSTANT_String:
+					break;
+				case CFR_CONSTANT_Class:
+					if (((flags & BCT_MajorClassFileVersionMask) < BCT_Java5MajorVersionShifted)
+					&& (J9_ARE_ANY_BITS_SET(flags, BCT_MajorClassFileVersionMask))
```
Comment (DanHeidinga) [inline runtime/bcverify/staticverify.c]: given the way this has been rewritten, it makes sense to just remove the `&& (J9_ARE_ANY_BITS_SET(flags, BCT_MajorClassFileVersionMask))` as I don't believe it's required anymore.

Previously, cfdump used it.

File: runtime/nls/cfre/cfrerr.nls [hunk 3]
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
 J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55.explanation=Please consult the Java Virtual Machine Specification for a detailed explaination.
 J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
 J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55.user_response=Contact the provider of the classfile for a corrected version.
+
+# END NON-TRANSLATABLE
+
+#ldc and ldc_w are not translatable
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC=The ldc and ldc_2 bytecodes must reference a constant or a constant dynamic
```
Comment (DanHeidinga) [inline runtime/nls/cfre/cfrerr.nls]: `constant dynamic`  -> `constant_dynamic`

File: runtime/nls/cfre/cfrerr.nls [hunk 4]
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
...
+# START NON-TRANSLATABLE
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.explanation=An 'ldc' or 'ldc_w' bytecode can only reference constant values or a constant dynamic entry.
```
Comment (DanHeidinga) [inline runtime/nls/cfre/cfrerr.nls]: same here

File: runtime/nls/cfre/cfrerr.nls [hunk 5]
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
...
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.user_response=Contact the provider of the classfile for a corrected version.
+
+# END NON-TRANSLATABLE
+
+#ldc2_w is not translatable
+J9NLS_CFR_ERR_BC_LDC2W_NOT_CONSTANT_OR_CONSTANT_DYNAMIC=The ldc2_w bytecode must reference a constant or a constant dynamic
```
Comment (DanHeidinga) [inline runtime/nls/cfre/cfrerr.nls]: same

File: runtime/nls/cfre/cfrerr.nls [hunk 6]
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
...
+# START NON-TRANSLATABLE
+J9NLS_CFR_ERR_BC_LDC2W_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.explanation=An 'ldc2_w' bytecode can only reference constant values or a constant dynamic entry.
+J9NLS_CFR_ERR_BC_LDC2W_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+J9NLS_CFR_ERR_BC_LDC2W_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.user_response=Contact the provider of the classfile for a corrected version.
+
+# END NON-TRANSLATABLE
+
+#ldc and ldc_w are not translatable
+J9NLS_CFR_ERR_BC_LDC_CONSTANT_DYNAMIC_RETURNS_LONG_OR_DOUBLE=Constant dynamic entries referenced by the ldc and ldc_w bytecodes must not return long or double
```
Comment (DanHeidinga) [inline runtime/nls/cfre/cfrerr.nls]: same.  Might be better to use Constant_Dynamic

File: runtime/nls/cfre/cfrerr.nls [hunk 7]
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
...
+# START NON-TRANSLATABLE
+J9NLS_CFR_ERR_BC_LDC_CONSTANT_DYNAMIC_RETURNS_LONG_OR_DOUBLE.explanation=An 'ldc' or 'ldc_w' bytecode can only reference constant values or a constant dynamic entry.
+J9NLS_CFR_ERR_BC_LDC_CONSTANT_DYNAMIC_RETURNS_LONG_OR_DOUBLE.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+J9NLS_CFR_ERR_BC_LDC_CONSTANT_DYNAMIC_RETURNS_LONG_OR_DOUBLE.user_response=Contact the provider of the classfile for a corrected version.
+
+# END NON-TRANSLATABLE
+
+#ldc2_w is not translatable
+J9NLS_CFR_ERR_BC_LDC2W_CONSTANT_DYNAMIC_NOT_LONG_OR_DOUBLE=Constant dynamic entries referenced by the ldc2_w bytecode must return long or double
```
Comment (DanHeidinga) [inline runtime/nls/cfre/cfrerr.nls]: same

**Comment by fengxue-IS** [comment]
removed the legacy check condition, changed nls to use `Constant_Dynamic`, also modified some error message wording to make the text more clear to users

**Comment by AdamBrousseau** [comment]
Jenkins line endings check

**Comment by DanHeidinga** [comment]
The extended test failures are the jdk11 issues we're still working through.  Not related to this PR

**Pull request merged**